typed_sql/lib/src/adaptor/postgres_adaptor.dart

```diff
@@ -0,0 +1,529 @@
+  }
+
+  @override
+  Future<void> close({bool force = false}) async {
+    _closing = true;
+
+    if (force) {
+      if (!_forceClosing) {
+        await _pool.close(force: true);
+      }
+      _forceClosing = true;
+    }
+
+    // Gracefully wait for all connection requests to be closed
+    while (_pendingConnectionRequests > 0) {
+      await _stateChanged.wait;
+    }
+
+    await _pool.close();
+
+    _closed = true;
+  }
+
+  @override
+  Stream<RowReader> query(String sql, List<Object?> params) async* {
+    try {
+      // We can't use _withConnection directly, because that would make it
+      // impossible to a streaming result.
+      final futureConn = Completer<Connection>();
+      final done = Completer<void>();
+      scheduleMicrotask(() async {
+        try {
+          await _withConnection((conn) {
+            futureConn.complete(conn);
+            return done.future;
+          });
+        } catch (e, st) {
+          futureConn.completeError(e, st);
+        }
+      });
+
+      try {
+        final conn = await futureConn.future;
+
+        // TODO: Explore why streaming mode doesn't work
+        // final ps = await _session.prepare(sql);
+        // yield* ps.bind(params).map(_PostgresRowReader.new);
+        final result = await conn.execute(
+          sql,
+          parameters: _paramsForPostgres(params),
+          queryMode: QueryMode.extended,
+        );
+        yield* Stream.fromIterable(result.map(_PostgresRowReader.new));
+      } finally {
+        done.complete();
+      }
+    } on Exception catch (e) {
+      _throwDatabaseException(e);
+    }
+  }
+
+  @override
+  Future<QueryResult> execute(String sql, List<Object?> params) async {
+    try {
+      return await _withConnection((conn) async {
+        final result = await conn.execute(
+          sql,
+          parameters: _paramsForPostgres(params),
+        );
+        return QueryResult(affectedRows: result.affectedRows);
+      });
+    } on Exception catch (e) {
+      _throwDatabaseException(e);
+    }
+  }
+
+  @override
+  Future<void> script(String sql) async {
+    try {
+      return await _withConnection((conn) async {
+        await conn.execute(sql, queryMode: QueryMode.simple);
+      });
+    } on Exception catch (e) {
+      _throwDatabaseException(e);
+    }
+  }
+
+  @override
+  Future<T> transact<T>(Future<T> Function(DatabaseTransaction tx) fn) async {
+    // We can't use _withConnection directly, because then we can't know where
+    // exceptions come from. We **must** know if an exception is from the
+    // database or application code, as these **must** be treated differently.
+    final futureConn = Completer<Connection>();
+    final done = Completer<void>();
+    scheduleMicrotask(() async {
+      try {
+        await _withConnection((conn) {
+          futureConn.complete(conn);
+          return done.future;
+        });
+      } catch (e, st) {
+        futureConn.completeError(e, st);
+      }
+    });
```

Comment:
It seems you are doing this two times... Would it be possible to abstract this into a helper, that establishes the connection in a microtask, and runs a callback? It could take error callbacks if needed...